perception/sub8_perception/nodes/buoy_finder.py

```diff
@@ -0,0 +1,364 @@
+            cx = int(M['m10'] / M['m00'])
+            cy = int(M['m01'] / M['m00'])
+            tpl_center = (int(cx), int(cy))
+            return cnt, tpl_center, area
+        else:
+            return None
+
+    def find_single_buoy(self, buoy_type):
+        '''
+        Attempt to find one color buoy in the image.
+        1) Create mask for buoy's color in colorspace specified in paramaters
+        2) Select the largest contour in this mask
+        3) Approximate a circle around this contour
+        4) Store the center of this circle and the current tf between /map and camera
+           as an observation
+        5) If observations for this buoy is now >= min_observations, approximate buoy
+           position using the least squares tool imported
+        '''
+        assert buoy_type in self.buoys.keys(), "Buoys_2d does not know buoy color: {}".format(buoy_type)
+        buoy = self.buoys[buoy_type]
+        mask = buoy.segment(self.last_image)
+        kernel = np.ones((5, 5), np.uint8)
+        mask = cv2.erode(mask, kernel, iterations=2)
+        mask = cv2.dilate(mask, kernel, iterations=2)
+
+        _, contours, _ = cv2.findContours(mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE,
+                                          offset=(self.roi[0], self.roi[1]))
+        ret = self.get_best_contour(contours)
+        if ret is None:
+            self.buoys[buoy_type].status = 'not seen'
+            return
+        contour, tuple_center, area = ret
+        true_center, rad = cv2.minEnclosingCircle(contour)
+
+        if self.debug_ros:
+            cv2.add(self.mask_image.copy(), self.buoys[buoy_type].cv_colors, mask=mask, dst=self.mask_image)
+            cv2.circle(self.mask_image, (int(true_center[0] - self.roi[0]), int(true_center[1]) - self.roi[1]),
+                       int(rad), self.buoys[buoy_type].cv_colors, 2)
+        if self.debug_cv:
+            self.debug_images[buoy_type] = mask.copy()
+
+        try:
+            self.tf_listener.waitForTransform('/map', self.frame_id, self.last_image_time, rospy.Duration(0.2))
+        except tf.Exception as e:
+            rospy.logwarn("Could not transform camera to map: {}".format(e))
+            return False
+
+        if not self.sanity_check(tuple_center, self.last_image_time):
+            self.buoys[buoy_type].status = 'failed sanity check'
+            return False
+
+        (t, rot_q) = self.tf_listener.lookupTransform('/map', self.frame_id, self.last_image_time)
+        trans = np.array(t)
+        R = mil_ros_tools.geometry_helpers.quaternion_matrix(rot_q)
+
+        if (buoy.last_t is None) or (np.linalg.norm(trans - buoy.last_t) > 0.1):
+            buoy.last_t = trans
+            buoy.observations.append(true_center)
+            buoy.pose_pairs.append((t, R))
+
+        if len(buoy.observations) > self.min_observations:
+            self.buoys[buoy_type].est = self.multi_obs.lst_sqr_intersection(buoy.observations, buoy.pose_pairs)
```

Comment:
I still don't think this a proper way to do this. The buoys do move under water. Last year they had a buoy attached to something on the surface and that one moved around like crazy. Even with truncating the length of the observations at 300, you'll keep 20 seconds or so of history. From my experience last year, the buoy could move a decent amount in that time.